src/GraphODataTemplateWriter/Extensions/OdcmModelExtensions.cs

```diff
@@ -292,6 +294,85 @@ public static OdcmProperty GetServiceCollectionNavigationPropertyForPropertyType
+                            ?.Where(singleton =>
+                                        singleton.
+                                                Type.AsOdcmClass()
+                                                ?.Properties
+                                                ?.Where(property => property.ContainsTarget) // find singleton type properties that are self contained 
+                                                ?.Any(property => property.IsPropertyTypeChainedContainedNavigationProperty( //  keep following the containsTarget=True to find if we can use a reference
+                                                                                odcmProperty,
+                                                                                navigatedTypes,
+                                                                                property.IsCollection
+                                                                                    ? $"{singleton.Name}/{property.Name}/{{id}}"
+                                                                                    : $"{singleton.Name}/{property.Name}")
+                                                ) ?? false
+                            ) ?? Array.Empty<OdcmProperty>()
+                    ).FirstOrDefault();
+
+            return matchingRootProperty != null;
+        }
+
+        /// <summary>
+        /// This method takes in a property and looks through to see if the given testProperty could be contained in it through its contained navigation properties or
+        /// its nested contained navigation properties.
+        /// </summary>
+        /// <param name="odcmRootProperty">The property to start the search from</param>
+        /// <param name="testProperty">The property to find in the from the root type</param>
+        /// <param name="navigatedTypes">The types already navigated to prevent circular references</param>
+        /// <param name="route">The route followed to get here</param>
+        /// <returns></returns>
+        private static bool IsPropertyTypeChainedContainedNavigationProperty(this OdcmProperty odcmRootProperty, OdcmProperty testProperty, HashSet<string> navigatedTypes, string route)
+        {
+            // check if the property already matches the type.
+            if (odcmRootProperty.Type.FullName.Equals(testProperty.Type.FullName))
+            {
+                logger.Info("Property \"{0}\" matches self contained navigation property \"{1}\" of type \"{2}\"", testProperty.Name, odcmRootProperty.Name, odcmRootProperty.Type.FullName);
+                logger.Info("Possible route from service class is: {0}{1}{1}",route,Environment.NewLine);
+                return true;
+            }
+
+            // check if the base class is referenced instead.
+            if (odcmRootProperty.Type.FullName.Equals(testProperty.BaseClass()?.FullName))
```

Comment:
Do we need to check whether any class in the base class chain is referenced instead of only the immediate base class?

Reply:
I agree with this point. 
So far, this case is never hit though with the v1 or beta metadata. I still need to test if there are any hits if we go further up the chain. If not, I might just consider dropping this condition

Reply:
Resolved via https://github.com/microsoftgraph/MSGraph-SDK-Code-Generator/pull/596/commits/b1ecd0cc1114194da30039997016377ccf27c789